A veg-modifier-model-run-prototype is created

ruiwu1990 opened the pull request:
1.  http://192.168.99.100:5010/scenario_table is the main page
2. The basic workflow is that users go to /scenario_table to create a scenario or view an existing scenario. If users choose to create one, then the system will go to http://192.168.99.100:5010/create_new_scenario, a list of model runs will be shown in a table. Users use a 2D map tool to modify veg variables; Give scenario a name; Run scenario; system redirect users to the main page and the new created scenario will be shown in the table.
3.  Users need to have at least a model run in Moinul’s server. This is because the workflow. This will cause a problem (have no model run to create a scenario), if you are a new users and do not have any model runs in Moinul server, you may go to http://192.168.99.100:5010 to create a model run.
4. I have done a small test that different scenarios are different: change all the HRU into bare ground vs shrub, the hydrographs are different
5. You may have a problem with authorization if you use Chrome instead of FireFox. I do not know why some times Chrome complains the API key (JWT authorize) expired but FireFox never does that.

TODO:
1. In http://192.168.99.100:5010/create_new_scenario, only display finished model run
2. Add authorization in the system.

Commit:
A prototype is done

Basic two things left:
1) model run with the corresponding files, need to modify api/scenarios
post request part
2) ajax get with the metadata—model_run_table.js

Commit:
change the work a little bit

Commit:
send metadata from get request

Commit:
Prototype is created

Commit:
Fixed a small thing

Commit:
Remove useless code

mt-digital commented on app/api/views.py:
```diff
@@ -78,12 +79,149 @@ def display_modelruns():
     """
     Handle get and push requests for list of all finished scenarios and submit
     a new scenario, respectively.
     """
```
please merge the latest master branch before opening pull requests

mt-digital commented on app/main/views.py:
```diff
@@ -11,9 +12,13 @@ def index():
+	return render_template('scenario_table.html')
+
+@main.route('/create_new_scenario')
+def create_new_scenario():
```
PEP8. Please fix this and any others

mt-digital commented on app/static/user_data/a@b.nc:
```diff
@@ -0,0 +1,389 @@
+GSFLOW Control File for Lehman Creek Chao Chen 
```
this file is not a netCDF file. Seems this was wrongly included in your commit

mt-digital commented on app/static/user_data/a@b/temp_control.control:
```diff
@@ -0,0 +1,389 @@
+GSFLOW Control File for Lehman Creek Chao Chen 
```
Same note here. This doesn't seem like it should actually be included

mtpain commented:
This isn't ready for a merge yet, there are quite a few problems. In general I need you to be more careful submitting pull requests: don't include errant data and please follow style standards.  Also make sure to merge the current master branch before opening pull requests.

Also, it appears you are developing with a Dockerized version of the system. Please include instructions both on how to run Docker and how to work with the code running in the Docker container.

Regarding 3)

```
Users need to have at least a model run in Moinul’s server. This is because the workflow. This will cause a problem (have no model run to create a scenario), if you are a new users and do not have any model runs in Moinul server, you may go to http://192.168.99.100:5010 to create a model run.
```

This won't work. There has to be a default, pre-initialized set of inputs to use. We have to enable new users to get going even if they don't have data ready.

Commit:
Fixed the PEP8. and add default model run files

TODO:
1. In http://192.168.99.100:5010/create_new_scenario, only display
finished model run
2. Add authorization in the system.
3. Enable users to run multiple scenario at the same time. This means
need to rewrite the post request ‘/api/scenarios’ in api/views.py

ruiwu1990 commented:
I tried to merge the latest master branch, but I found you have deleted 'PRMSCoverageTool.py'. I have modified this file and use some functions in this file. Can you let me use this file for current version? 3. is fixed. Users do not need to have inital model run.

mtpain commented:
The scenario tool doesn't seem to be working for me. I followed your instructions and tried to create a new scenario.

When I navigate to `/create_new_scenario` I see the screen below:

<img width="500" alt="screenshot 2016-05-04 10 34 55" src="https://cloud.githubusercontent.com/assets/2425472/15023240/8abc2088-11e4-11e6-8ce4-f7b3fcf7c809.png">

However in the console there is an error. Here's the traceback:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/usr/local/lib/python2.7/dist-packages/flask_cors/extension.py", line 188, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask_cors/extension.py", line 188, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/www/prms-veg/app/api/views.py", line 116, in scenarios
    BASE_PARAMETER_NC = input_file_folder + app.config['TEMP_PARAM']
KeyError: 'TEMP_PARAM'
```

When I click the "Choose Me"  nothing happens and there is another error in the console. Here is the request and traceback:

```
192.168.99.1 - - [04/May/2016 17:41:09] "GET /api/scenarios/download_input_files/http:++++++192.168.99.100:5000+++api+++modelresources+++download+++lc.control---http:++++++192.168.99.100:5000+++api+++modelresources+++download+++lc-data.nc---http:++++++192.168.99.100:5000+++api+++modelresources+++download+++grassy-knoll-tester.nc HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/usr/local/lib/python2.7/dist-packages/flask_cors/extension.py", line 188, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask_cors/extension.py", line 188, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/www/prms-veg/app/api/views.py", line 99, in download_model_inputs
    download_prms_inputs(control_url, data_url, param_url)
  File "/var/www/prms-veg/app/api/util.py", line 151, in download_prms_inputs
    os.mkdir(app_root)
OSError: [Errno 2] No such file or directory: '/var/www/prms-veg/app/api/../static/user_data/maturner01@gmail'
```

Commit:
add temperature modification

Commit:
temperature modification and create model run

class Scenario and ScenarioRun should be modified for temperature
modification

Commit:
small changes for time

Commit:
a little design change

Commit:
enable users to download hydro csv

The csv file is sent direct to the client without creating a temp file
in the server. I planned to create a checkbox for each scenario and
enable users to download all the data as a csv file but I found each
scenario can have different timestamps or even different length (e.g.
s1 is for one year and s2 is for ten years). It is not very meaningful
for me to combine all the scenarios together in a single csv file.

Commit:
remove unnecessary files

Commit:
initialize the session settings

Commit:
session works, need to keep working on token rewrite

Commit:
finalize dockerize

Commit:
another try to update python in docker container

Commit:
questionable token

Matthew’s latest ModelApiClient does not work for me. There is a bug:
5/20/2016 2:36:41 PMTraceback (most recent call last):
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1836, in
__call__
5/20/2016 2:36:41 PM    return self.wsgi_app(environ, start_response)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1820, in
wsgi_app
5/20/2016 2:36:41 PM    response =
self.make_response(self.handle_exception(e))
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask_cors/extension.py", line
188, in wrapped_function
5/20/2016 2:36:41 PM    return
cors_after_request(app.make_response(f(*args, **kwargs)))
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1403, in
handle_exception
5/20/2016 2:36:41 PM    reraise(exc_type, exc_value, tb)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1817, in
wsgi_app
5/20/2016 2:36:41 PM    response = self.full_dispatch_request()
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1477, in
full_dispatch_request
5/20/2016 2:36:41 PM    rv = self.handle_user_exception(e)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask_cors/extension.py", line
188, in wrapped_function
5/20/2016 2:36:41 PM    return
cors_after_request(app.make_response(f(*args, **kwargs)))
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1381, in
handle_user_exception
5/20/2016 2:36:41 PM    reraise(exc_type, exc_value, tb)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1475, in
full_dispatch_request
5/20/2016 2:36:41 PM    rv = self.dispatch_request()
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/flask/app.py", line 1461, in
dispatch_request
5/20/2016 2:36:41 PM    return
self.view_functions[rule.endpoint](**req.view_args)
5/20/2016 2:36:41 PM  File "/var/www/prms-veg/app/api/views.py", line
229, in scenarios
5/20/2016 2:36:41 PM    model_host=app.config['MODEL_HOST']
5/20/2016 2:36:41 PM  File
"/var/www/prms-veg/app/api/PRMSCoverageTool.py", line 130, in run
5/20/2016 2:36:41 PM    modelrun=dict(title=self.scenario_name,
model_name='prms')
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/client/swagger_client/apis/defau
lt_api.py", line 119, in create_modelrun
5/20/2016 2:36:41 PM    callback=params.get('callback'))
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/client/swagger_client/api_client
.py", line 322, in call_api
5/20/2016 2:36:41 PM    response_type, auth_settings, callback)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/client/swagger_client/api_client
.py", line 149, in __call_api
5/20/2016 2:36:41 PM    post_params=post_params, body=body)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/client/swagger_client/api_client
.py", line 358, in request
5/20/2016 2:36:41 PM    body=body)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/client/swagger_client/rest.py",
line 213, in POST
5/20/2016 2:36:41 PM    body=body)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/client/swagger_client/rest.py",
line 147, in request
5/20/2016 2:36:41 PM    headers=headers)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/request.py", line 73,
in request
5/20/2016 2:36:41 PM    **urlopen_kw)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/request.py", line 151,
in request_encode_body
5/20/2016 2:36:41 PM    return self.urlopen(method, url, **extra_kw)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/poolmanager.py", line
165, in urlopen
5/20/2016 2:36:41 PM    response = conn.urlopen(method, u.request_uri,
**kw)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/connectionpool.py",
line 566, in urlopen
5/20/2016 2:36:41 PM    conn = self._get_conn(timeout=pool_timeout)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/connectionpool.py",
line 256, in _get_conn
5/20/2016 2:36:41 PM    return conn or self._new_conn()
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/connectionpool.py",
line 217, in _new_conn
5/20/2016 2:36:41 PM    strict=self.strict, **self.conn_kw)
5/20/2016 2:36:41 PM  File
"/usr/local/lib/python2.7/site-packages/urllib3/connection.py", line
126, in __init__
5/20/2016 2:36:41 PM    _HTTPConnection.__init__(self, *args, **kw)
5/20/2016 2:36:41 PMTypeError: __init__() got an unexpected keyword
argument 'assert_hostname'

Commit:
dockerize

Commit:
hope final version

Commit:
tested works, but slow for NMU servers

Commit:
kind of hard coded for justin input

it is in model_run_table.js

Commit:
Justin's nc is transposed
